### ISSUE TITLE
Carefully formatted columns (matrix literal example)

### ISSUE THREAD

**Issue by anderejd** [rust-lang/rustfmt#2924]
Here's one interesting really bad case I just found:

```diff
             self.projection_matrix = Matrix4::new(
-                1.0/r, 0.0,   0.0,         0.0,   // NOTE: first column!
-                0.0,   1.0/t, 0.0,         0.0,   // 2nd
-                0.0,   0.0,   2.0/(n-f),   0.0,   // 3rd
-                0.0,   0.0,   (f+n)/(n-f), 1.0    // 4th
+                1.0 / r,
+                0.0,
+                0.0,
+                0.0, // NOTE: first column!
+                0.0,
+                1.0 / t,
+                0.0,
+                0.0, // 2nd
+                0.0,
+                0.0,
+                2.0 / (n - f),
+                0.0, // 3rd
+                0.0,
+                0.0,
+                (f + n) / (n - f),
+                1.0, // 4th
             );
```

Would it be possible to avoid breaking the formatting on a few select blocks? Maybe by some magic comment or by looking for lines that end with comments like this example?

**Comment by anderejd** [comment]
Would it make sense to never reformat lines that both ends with a `//` comment and has non-space characters before the comment starts?

If no, would config option to activate that behavior be ok?

**Comment by anderejd** [comment]
For anyone like me, not always reading the manual properly for your tools, see this workaround: 
https://github.com/rust-lang-nursery/rustfmt/blob/af3c3c694b842e047128e870f9bc8d56c7057c59/README.md#tips `#[cfg_attr(rustfmt, rustfmt::skip)]`

**Comment by nrc** [comment]
`#[rustfmt::skip]` has been stabilised. It works on nightly and will be riding the trains to stable.

**Comment by anderejd** [comment]
> will be riding the trains to stable.

Got a guess on ETA?

**Comment by topecongiro** [comment]
P-low because this is a good target for `rustfmt::skip`. In the long run, rustfmt may be able to use the position of comment to figure out the optimal layout (this is not guaranteed to happen, as in general we try to avoid relying on user's input).

**Comment by nilgoyette** [comment]
Thank you, this is now fixed with `#[rustfmt::skip]`.

However, I have a follow-up question concerning this "matrix" problem. We have dozens of matrix declarations in our project. Adding `#[rustfmt::skip]` does work, but I must add a lof of `#[rustfmt::skip]` in several files. Is there a tool to ask to skip the formatting for a list of objects in the whole project or for a folder? Something like

    #[rustfmt::skip(Matrix3, Matrix4)]

in lib.rs or in a mod.rs?

**Comment by nilgoyette** [comment]
Thank you for the information, it might serve someday. It's not really what I'm asking though. It's more: skiping on a per-struct basis.

**Comment by anderejd** [comment]
Is it solved now?

**Comment by nilgoyette** [comment]
@anderejd Reading your original question "_Maybe by some magic comment [...]_", yes, this is solved. You can use `skip` like this:

    #[rustfmt::skip]
    let m = Matrix3::new(
        1.0, 0.5, 0.0,
        0.0, 1.0, 0.5,
        0.0, 0.5, 1.0,
    );

and rustfmt won't touch anything until the line finishes, so the last `)`. You can also skip the formatting for an entire module. Sadly, there's no way to skip the formatting for all `[Matrix3, Matrix4, ...]` in a project though.

**Comment by anderejd** [comment]
IMHO it's not fixed, but I am thankful for the workaround.

**Comment by tgross35** [comment]
I did not see this issue so I wrote #5553 (closed as duplicate) but would like to reiterate that there is support for an option like this. A [reddit comment](https://www.reddit.com/r/rust/comments/xreg1a/comment/iqeur3y) chain got 66 upvotes on a comment about rustfmt not being aware of matrices (as expected), and the response suggesting `#[rustfmt::items(10)]` got 33 upvotes (of course many of these users may be people who never use the features, but these are what inspired me to write #5553)

If it's not a hard no, I would request reopening this issue to help the math community. `#[rustfmt::skip]` is "acceptable", but a builtin alternative to hand formatting large arrays would be quite nice for everybody.
